src/api/app/models/workflow/step/branch_package_step.rb

```diff
@@ -0,0 +1,77 @@
+class Workflow
+  module Step
+    class BranchPackageStep
+      include ActiveModel::Model
+
+      attr_reader :errors
+
+      validates :source_project, :source_package, presence: true
+
+      def initialize(step_instructions:, pr_number:)
+        @step_instructions = step_instructions
+        @pr_number = pr_number
+        @errors = []
+      end
+
+      def call
+        branch
+      end
+
+      private
+
+      def source_project
+        @step_instructions['source_project']
+      end
+
+      def source_package
+        @step_instructions['source_package']
+      end
+
+      def destination_package
+        return @step_instructions['destination_package'] if @step_instructions['destination_package'].present?
+
+        source_package
+      end
+
+      def destination_project
+        source_project + ":PR-#{@pr_number}"
+      end
+
+      def remote_source?
+        return true if Project.find_remote_project(source_project)
+
+        false
+      end
+
+      def check_source_access
+        return if remote_source?
+
+        options = { use_source: false, follow_project_links: true, follow_multibuild: true }
+        src_package = Package.get_by_project_and_name(source_project, source_package, options)
+
+        raise Pundit::NotAuthorizedError unless PackagePolicy.new(User.session, src_package).create_branch?
```

Comment:
```suggestion
        authorize src_package, :create_branch?
```

Reply:
@krauselukas and I talked about this last week and `authorize` isn't available outside controllers. I just found out by looking at the Pundit code that `Pundit.authorize` is exactly the same, but also available outside controllers. With this in mind, I would suggest doing this:

```suggestion
        Pundit.authorize src_package, :create_branch?
```

Reply:
`Pundit.authorize` is just `authorize` but with `include Pundit` being called before, no? In another hand is that a good thing to leak Pundit authorization to models?

Reply:
I don't see how we can do authorization in the controller for this since we need to parse the YAML file, then check if the token user is authorized or not to do whatever is defined in the YAML file.